Implement ability to fall back to secondary Eth1 nodes if the primary node is unavailable

AbdelStark opened the pull request:
## PR Description
As a staker, I want to be able to fail-over automatically to a secondary/tertiary Eth1 node so that if the primary Eth1 node is unavailable or incorrect I am able to continue to propose blocks correctly.

## Changes summary

Introduce a fallback mechanism to select the best `Eth1Provider`. 

- Added `FallbackAwareEth1Provider` that implements `Eth1Provider`
  - Takes a list of `Eth1Provider` corresponding to the candidates
  - Instantiate an `Eth1ProviderSelector` 
- Updated `--eth1-endpoint` CLI option
  - Added `--eth1-endpoints` alias for the option
  - Updated to a list of endpoints
- Updated `PowchainService` to build the `Eth1Provider` accordingly
  - If only one endpoint is specified a classical provider is built
  - If more than one endpoint is specified  a `FallbackAwareEth1Provider` is built

## Fixed Issue(s)
Addresses #3327 

## Documentation

- [X] I thought about documentation and added the `documentation` label to this PR if updates are required.

AbdelStark committed:
sonar support

AbdelStark committed:
Merge pull request #1 from abdelhamidbakhta/sonar

sonar support

AbdelStark committed:
Merge remote-tracking branch 'upstream/master'

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

AbdelStark committed:
Merge remote-tracking branch 'upstream/master'

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

AbdelStark committed:
Merge remote-tracking branch 'upstream/master'

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

AbdelStark committed:
Introduce a fallback mechanism to select the best Eth1Provider

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

AbdelStark committed:
remove sonar changes

AbdelStark committed:
Update RoundRobinEth1ProviderSelector.java

ajsutton commented on build.gradle:
```diff
@@ -22,24 +22,24 @@ apply plugin: 'application'
-    'build',
-    'test']
-  ]
+        'dev': [
```
Would be good to avoid this reformatting in this PR.

AbdelStark replied:
done. reset to master branch

ajsutton commented on pow/src/main/java/tech/pegasys/teku/pow/fallback/readiness/Eth1ProviderChainHeadReadiness.java:
```diff
@@ -0,0 +1,49 @@
+    } catch (InterruptedException | ExecutionException | TimeoutException e) {
+      return false;
+    }
+  }
```
Not sure making a test request is the right way to check for readiness generally.  I think it makes sense to periodically poll the nodes to check a) they have the right chain ID and b) that they are in sync and consider the node not ready if either of those aren't true (or if requests to get that data fail).

Otherwise, we'd just try each request with the first ready node, and if it doesn't work, retry the request with the next ready node in the list.  We'd consider a request as "not working" if the request completely failed or timed out, if we requested a block but didn't get data (node not yet finished downloading historic data) or if we got a response that skipped some deposit logs (node hasn't finished downloading historic logs).  

Handling out of order deposits might need to be a follow up PR though as it could be that an earlier request didn't return all deposits earlier in the chain and we need to actually retry with a different block range.  That probably can't be done all with in the Eth1Provider and may not be worth the complexity.

AbdelStark replied:
ok that makes sense. i will update the implementation accordingly

AbdelStark replied:
i have simplified the implementation. basically if a call to eth1 provider fails, the next call will select the next candidate based on the list of providers.

ajsutton commented on pow/src/main/java/tech/pegasys/teku/pow/fallback/strategy/FallbackStrategy.java:
```diff
@@ -0,0 +1,20 @@
+public enum FallbackStrategy {
+  PRIORITY,
+  RANDOM,
+  ROUND_ROBIN
```
I don't see a need for anything beyond priority here.  The main use-case will be people running their own Eth1 node but wanting to automatically fallback to infer if it's not yet in sync or goes down.

AbdelStark replied:
Ok i will remove that

ajsutton reviewed (commented):
Thanks for picking this up - would definitely be really cool to have this support.  I'd aim to simplify this a fair bit - if people want complex load balancing setups they can do that with an external load balancer and we don't want to reimplement all that logic.  But basic support for priority order fallback if a request fails or the node isn't yet ready would be awesome to have built in to make things easier for people.

AbdelStark committed:
Merge remote-tracking branch 'upstream/master' into feature/3327-fallback-eth1-nodes

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

AbdelStark committed:
address PR comments. mainly simplify the approach. swith eth1 provider for next call if current call failed.

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

AbdelStark committed:
Merge remote-tracking branch 'upstream/master' into feature/3327-fallback-eth1-nodes

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

AbdelStark committed:
Update build.gradle

AbdelStark committed:
dont ignore returned future

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

AbdelStark committed:
add missing override annotation on method

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

AbdelStark committed:
suppress warnings in test class

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

ajsutton commented on pow/src/main/java/tech/pegasys/teku/pow/fallback/FallbackAwareEth1Provider.java:
```diff
@@ -0,0 +1,97 @@
+              "Error caught while calling Eth1Provider, switching provider for next calls",
+              throwable);
+          eth1ProviderSelector.updateBestCandidate();
+        });
```
I think this should be retrying the call on the next provider.  If you have a local node and infura configured, this will cause the first failure in your local node to trigger a fail over to infer and since infura won't go down it will never switch back.

I'd also actually do the retry with the next "ready" provider inside this class - that way the caller only gets an error if all providers are unavailable.

It would probably make sense to mark a provider as offline/unhealthy for some period if we get a failure from it to avoid repeatedly retrying calls to it before moving on to the next provider.  Might be a followup PR as it would work in really nicely with a periodic in-sync check - you'd mark a provider as offline if a request failed and mark it online again when the in-sync check next passes.

AbdelStark replied:
Done. The retry is done directly inside this class.

ajsutton commented on services/powchain/src/main/java/tech/pegasys/teku/services/powchain/PowchainService.java:
```diff
@@ -62,11 +66,19 @@ public PowchainService(final ServiceConfig config) {
+    final List<Eth1Provider> eth1Providers =
+        Arrays.asList(
+            new ThrottlingEth1Provider(
+                new ErrorTrackingEth1Provider(
```
We will probably have to adjust where the `ErrorTrackingEth1Provider` is put in this chain.  It may wind up needing to be aware of the multiple providers so it can provide a warning that the primary eth1 node is unavailable but backup is working vs all providers being down etc.  Possibly the functionality just moves into `FallbackAwareEth1Provider` - not sure exactly.

AbdelStark replied:
Done. The `ErrorTrackingEth1Provider` now encapsulates the `FallbackAwareEth1Provider`

ajsutton reviewed (commented):
Definitely much simpler which is great.  In terms of the functionality I think the key is just really focussing on the idea of having a local preferred node which may be unreliable and then a more centralised but highly reliable backup (e.g. infura). To aid decentralisation we want to use the first (local) node as much as possible and only fail over when it's not working, then switch back to it once it's available again.

AbdelStark committed:
Merge remote-tracking branch 'upstream/master' into feature/3327-fallback-eth1-nodes

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

# Conflicts:
#	services/powchain/src/main/java/tech/pegasys/teku/services/powchain/PowchainService.java

AbdelStark committed:
Address PR comments
Do the retry logic directly instead of waiting the next calls.
If a provider fails we test other providers directly.

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

AbdelStark committed:
Fix unit tests

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

ajsutton commented on pow/src/main/java/tech/pegasys/teku/pow/fallback/strategy/PriorityEth1ProviderSelector.java:
```diff
@@ -0,0 +1,51 @@
+
+  @SuppressWarnings("FutureReturnValueIgnored")
+  public PriorityEth1ProviderSelector(final List<Eth1Provider> candidates) {
+    assert candidates != null && !candidates.isEmpty();
```
nit: Should use `Predicates.checkArgument` rather than the `assert` keyword here (`assert` keyword is skipped by default which gives a false sense of security).

AbdelStark replied:
Done

ajsutton commented on pow/src/main/java/tech/pegasys/teku/pow/fallback/strategy/PriorityEth1ProviderSelector.java:
```diff
@@ -0,0 +1,51 @@
+    int nextIndex = (bestCandidateIndex.get() + 1) % candidates.size();
+    bestCandidateIndex.set(nextIndex);
+    bestCandidate.set(candidates.get(bestCandidateIndex.get()));
+  }
```
This is still setup so that it will get "stuck" using a backup node.  For example if I'm using a local Eth1 node first and then Infura second, the first time my Eth1 node returns an error, the best candidate will be updated to use Infura then because Infura is reliable it is unlikely to get any further errors and will just always use Infura but we actually want it to switch back to using the local node as soon as its available again.

I would probably drop the concept of best candidate and just have this return a list of available providers (in follow up PRs it could filter them based on whether its in-sync or not, on the right chain or has returned errors recently etc).  That would also fix the race condition in `FallbackAwareEth1Provider` if it was just iterating through the list.

AbdelStark replied:
I am now using an iterator to iterate over the list of available providers, restarting always with the first.

AbdelStark replied:
I dropped the concept of best candidate

ajsutton commented on pow/src/main/java/tech/pegasys/teku/pow/fallback/FallbackAwareEth1Provider.java:
```diff
@@ -0,0 +1,109 @@
+                      } else {
+                        return SafeFuture.failedFuture(err);
+                      }
+                    }));
```
There's a race condition here where if two requests are in progress at the same time, they'll both be updating the best candidate and so wind up potentially iterating through the list multiple times.  I think it works out ok but the behaviour is pretty hard to reason about.

AbdelStark replied:
Ok, i am now using an iterator as you suggested.

ajsutton commented on pow/src/test/java/tech/pegasys/teku/pow/fallback/strategy/PriorityEth1ProviderSelectorTest.java:
```diff
@@ -0,0 +1,73 @@
+    return blockSafeFuture;
+  }
+
+  private static SafeFuture<EthBlock.Block> koProvider() {
```
nit: not sure what `ko` means here. Maybe just `failingProvider`?

AbdelStark replied:
Done

ajsutton commented on pow/src/main/java/tech/pegasys/teku/pow/fallback/strategy/Eth1ProviderSelector.java:
```diff
@@ -0,0 +1,25 @@
+
+import tech.pegasys.teku.pow.Eth1Provider;
+
+public interface Eth1ProviderSelector {
```
nit: I think I'd scrap this interface since we only have one implementation.  We can always extract an interface later if we wind up with multiple implementations.

AbdelStark replied:
Ok. I removed the interface

AbdelStark committed:
Merge remote-tracking branch 'upstream/master' into feature/3327-fallback-eth1-nodes

AbdelStark committed:
Drop the concept of best candidate and use and iterator to fix the race condition

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

ajsutton commented on teku/src/main/java/tech/pegasys/teku/cli/options/DepositOptions.java:
```diff
@@ -27,11 +28,11 @@
-      description = "URL for Eth1 node.",
-      arity = "1")
-  private String eth1Endpoint = null;
+      description = "URL for Eth1 nodes.",
```
So that `--eth1-endpoint http://localhost:8545,http://oopsy.local:8545` works:

```suggestion
      description = "URL for Eth1 nodes.",
      split = ",",
```

ajsutton replied:
Actually, given that we haven't got the `DepositContract` updated yet to support fallback, that makes it pretty much impossible to test this well, so I think we're probably better off leaving the CLI as accepting only a single argument at this point until we get the internals fully sorted.  Basically toggle off the feature at the CLI level for the moment.  So this field would be a `String` and not a `List` but then `getEth1Endpoints()` method would still return a `List` - it would just never contain more than one item.

We could add a new, hidden `--eth1-endpoints` option to make testing easier if you wanted to and then merge the two options when we're ready to support it, but could also just modify the code for local testing for now.

Sorry that mucks you around again - just that this is a fairly popular feature so I suspect people will discover it if its exposed and that will create confusion.

ajsutton commented on services/powchain/src/main/java/tech/pegasys/teku/services/powchain/PowchainService.java:
```diff
@@ -149,10 +175,16 @@ private static OkHttpClient createOkHttpClient() {
+        web3js.stream()
+            .map(web3j -> SafeFuture.fromRunnable(web3j::shutdown))
+            .collect(Collectors.toList());
+    stopServices.addAll(shutdownWeb3js);
```
Because `Arrays.asList` returns an immutable list, this throws:
```
12:06:55.357 ERROR - PLEASE FIX OR REPORT | Unexpected exception thrown for Thread-5
java.lang.UnsupportedOperationException: null
	at java.util.AbstractList.add(AbstractList.java:153) ~[?:?]
	at java.util.AbstractList.add(AbstractList.java:111) ~[?:?]
	at java.util.AbstractCollection.addAll(AbstractCollection.java:336) ~[?:?]
	at tech.pegasys.teku.services.powchain.PowchainService.doStop(PowchainService.java:187) ~[teku-services-powchain-develop.jar:20.12.1+28-g010912a]
	at tech.pegasys.teku.service.serviceutils.Service.stop(Service.java:40) ~[teku-services-serviceutils-develop.jar:20.12.1+28-g010912a]
	at java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195) ~[?:?]
	at java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1625) ~[?:?]
	at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:484) ~[?:?]
	at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:550) ~[?:?]
	at java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:260) ~[?:?]
	at java.util.stream.ReferencePipeline.toArray(ReferencePipeline.java:517) ~[?:?]
	at tech.pegasys.teku.services.ServiceController.doStop(ServiceController.java:39) ~[teku-develop.jar:20.12.1+28-g010912a]
	at tech.pegasys.teku.service.serviceutils.Service.stop(Service.java:40) ~[teku-services-serviceutils-develop.jar:20.12.1+28-g010912a]
	at tech.pegasys.teku.BeaconNode.stop(BeaconNode.java:107) ~[teku-develop.jar:20.12.1+28-g010912a]
	at tech.pegasys.teku.Teku.lambda$start$0(Teku.java:53) ~[teku-develop.jar:20.12.1+28-g010912a]
	at java.lang.Thread.run(Thread.java:832) [?:?]
```

Not sure if this is getting too clever but you could replace the whole method with:

```
    return SafeFuture.allOfFailFast(
        Stream.concat(
                Stream.<ExceptionThrowingRunnable>of(
                    headTracker::stop, eth1DepositManager::stop, chainIdValidator::stop),
                web3js.stream().map(web3j -> web3j::shutdown))
            .map(SafeFuture::fromRunnable)
            .toArray(SafeFuture[]::new));
```

ajsutton commented on services/powchain/src/main/java/tech/pegasys/teku/services/powchain/PowchainService.java:
```diff
@@ -119,9 +128,26 @@ public PowchainService(final ServiceConfig config) {
+  }
+
+  private Web3j createWeb3j(final String endpoint) {
+    final HttpService web3jService = new HttpService(endpoint, createOkHttpClient());
```
Missed this before, but we should only use a single OkHttpClient instance that's shared among all the `HttpService` instances we create.

ajsutton commented on pow/src/main/java/tech/pegasys/teku/pow/fallback/FallbackAwareEth1Provider.java:
```diff
@@ -0,0 +1,109 @@
+  @Override
+  public SafeFuture<Optional<Block>> getEth1BlockWithRetry(
+      final UInt64 blockNumber, final Duration retryDelay, final int maxRetries) {
+    return run(
+        eth1Provider -> eth1Provider.getEth1BlockWithRetry(blockNumber, retryDelay, maxRetries));
```
These calls that automatically retry need to be handled a bit smarter - because `Web3jEth1Provider.getEth1BlockWithRetry` and the `getGuaranteedEth1Block` methods automatically retry, we wind up delegating the call to the first Eth1Provider which continuously retries even though its down and we never move on to the other providers.  I think we need to have these methods delegate to the plain `getEth1Block` but then handle the retry logic here - particularly for `getGuaranteedEth1Block` it just keeps retrying until it eventually gets the block, whereas `getEth1BlockWithRetry` retries a certain number of times then gives up.

ajsutton reviewed (commented):
Sorry, missed a few things earlier and only noticed them now that I'm testing the whole thing and seeing how it really runs.  The log output is really very verbose at the moment and makes it pretty hard to see exactly what's happening but it may be better to tidy that up in a follow up PR after we get the deposit contract supporting fallback.
